orderbook/src/api/order_validation.rs

```diff
@@ -0,0 +1,734 @@
+}
+
+impl WarpReplyConverting for ValidationError {
+    fn to_warp_reply(&self) -> (Json, StatusCode) {
+        match self {
+            ValidationError::Pre(pre) => pre.to_warp_reply(),
+            ValidationError::Post(post) => post.to_warp_reply(),
+        }
+    }
+}
+
+impl WarpReplyConverting for PostValidationError {
+    fn to_warp_reply(&self) -> (Json, StatusCode) {
+        match self {
+            Self::UnsupportedToken(token) => (
+                super::error("UnsupportedToken", format!("Token address {}", token)),
+                StatusCode::BAD_REQUEST,
+            ),
+            Self::WrongOwner(owner) => (
+                super::error(
+                    "WrongOwner",
+                    format!(
+                        "Address recovered from signature {} does not match from address",
+                        owner
+                    ),
+                ),
+                StatusCode::BAD_REQUEST,
+            ),
+            Self::InsufficientFunds => (
+                super::error(
+                    "InsufficientFunds",
+                    "order owner must have funds worth at least x in his account",
+                ),
+                StatusCode::BAD_REQUEST,
+            ),
+            Self::InsufficientFee => (
+                super::error("InsufficientFee", "Order does not include sufficient fee"),
+                StatusCode::BAD_REQUEST,
+            ),
+            Self::ZeroAmount => (
+                super::error("ZeroAmount", "Buy or sell amount is zero."),
+                StatusCode::BAD_REQUEST,
+            ),
+            Self::Other(_) => (super::internal_error(), StatusCode::INTERNAL_SERVER_ERROR),
+        }
+    }
+}
+
+/// Order validator plays the role of order validation over two phases.
+/// 1. Pre-Validation:
+///     This validation is aimed at catching malformed order data during the fee & quote phase
+///     before the order is signed. Thus, pre-validation doesn't verify signatures,
+///     user sell balances or fee sufficiency.
+///     Specifically, pre-validation verifies
+///     - if buy token is native asset, receiver is not a smart contract,
+///     - the sender is not a banned user,
+///     - the order validity is appropriate,
+///     - buy_token is not the same as sell_token,
+///     - buy and sell token destination and source are supported.
+/// 2. Post-Validation:
+///     This is the additional validation performed at the time of order placement
+///     (i.e. once all the required fields on an Order are provided). Specifically, verifying that
+///     - buy & sell amounts are non-zero,
+///     - order's owner matches the from field (if specified),
+///     - fee is sufficient,
+///     - buy & sell tokens passed "bad token" detection,
+///     - user has sufficient (transferable) funds to execute the order.
+///
+///     Furthermore, Post-order validation also calls pre-validate to ensure that
+///     other aspects of the order are not malformed.
```

Comment:
I worry that this description might quickly become out of date. Maybe a higher-level description on each of the methods instead of the type might make it closer to the actual code and less likely to become outdated over time.